lib/engine/game/g_18_india/step/corporate_buy_shares_company.rb

```diff
@@ -0,0 +1,183 @@
+# frozen_string_literal: true
+
+require_relative '../../../step/corporate_buy_shares'
+require_relative '../../../step/share_buying'
+require_relative '../../../action/corporate_buy_shares'
+require_relative '../../../action/corporate_buy_company'
+
+module Engine
+  module Game
+    module G18India
+      module Step
+        class CorporateBuySharesCompany < Engine::Step::CorporateBuyShares
+          include Engine::Step::ShareBuying
+
+          PURCHASE_ACTIONS = [Action::CorporateBuyShares, Action::CorporateBuyCompany].freeze
+
+          # used for debugging only
+          def setup
+            entity = current_entity
+            @log << 'Setup'
+            @log << "> step #{inspect}"
+            @log << "> entity: #{entity.name}"
+            @log << "> num_certs: #{@game.num_certs(entity)} / cert_limit: #{@game.cert_limit(entity)}"
+            @log << "> corporations_bought: #{@round.corporations_bought[entity]}"
+          end
+
+          def debugging_log(str)
+            @log << str
+            @log << ">  Num Certs: #{@game.num_certs(current_entity)} / Cert Limit: #{@game.cert_limit(current_entity)}"
+            @log << ">  Bought?: #{bought?(current_entity)} - Last: #{last_bought(current_entity).name} "
+          end
+
+          # update description
+          def description
+            'Corporate Purchase Certificates from IPO or Market'
+          end
+
+          # TODO: Convert Royal Bond in Phase IV
+          def actions(entity)
+            return [] if entity.nil? || !entity&.corporation?
+            return [] unless entity == current_entity
+
+            # May buy any share from Market (it may may itself from Market) [working]
+            # May buy Private Company from Market / Bank [working]
+            # May buy Railroad Bond from Market / Bank (doesn't count against Cert Limit of 3) [working]
+            # may buy One Proxy Cert from the IPO Row (but NOT itself) [working]
+            # convert a Royal Bond to a share of GIPR (Phase IV && must have space) [TODO]
+            actions = []
+            actions << 'corporate_buy_company' if can_buy_any_companies?(entity)
+            actions << 'corporate_buy_shares' if can_buy_any?(entity)
+            actions << 'pass' if actions.any?
+
+            actions
+          end
+
+          # ----- methods for buying companies (new) -----
+
+          def can_buy_any_companies?(entity)
+            return false if bought?(entity)
+
+            buyable_companies(entity).count.positive?
+          end
+
+          def buyable_companies(entity)
+            return [] unless entity.corporation?
+
+            companies = []
+            companies += @game.bank_owned_companies.select { |c| can_buy_comp_from_market?(entity, c) }
+            companies += @game.top_of_ipo_rows.select { |c| can_buy_comp_from_ipo?(entity, c) }
+
+            companies
```

Comment:
`+=` reallocates the `companies` array every time, you could use [`concat`](https://docs.ruby-lang.org/en/3.2/Array.html#method-i-concat) instead, or just add once:

```suggestion
            @game.bank_owned_companies.select { |c| can_buy_comp_from_market?(entity, c) } +
              @game.top_of_ipo_rows.select { |c| can_buy_comp_from_ipo?(entity, c) }
```